feat(rf/stack): add new resource support (#2551)
huaweicloud/terraform-provider-huaweicloud: docs/resources/rf_stack.md

---

> subcategory: "Application Orchestration Service (AOS)"

Author: As the category for provider document, the name is `Application Orchestration Service (AOS)`.
And for others, `RF`. is used.

---

# huaweicloud_rf_stack

> Manage a RF resource stack and deployment resources through this resource.

Reviewer: Manage an RF resource stack through this.

Author: done.

> Manage an RF resource stack and deployment resources through this resource.

Reviewer: This seems a bit difficult for understading, how about using something short like "Provides an RF resource stack."

Author: sounds great.

## Example Usage

> ### Create a RF resource stack and deployment VPC resources

Reviewer: Create an RF resource stack with VPC deployment

Author: done.

Reviewer: seems not done...

> ### Create an RF resource stack and deployment VPC resources

Reviewer: Create an RF resource stack with VPC deployment

Author: done.

```HCL
variable "stack_name" {}
variable "agency_name" {}
variable "region_name" {}
variable "vpc_name" {}
variable "subnet_name" {}

resource "huaweicloud_rf_stack" "test" {

>   name = var.instance_id

Reviewer: instance_id?

Author: `stack_name` is right.

  "provider": {
    "huaweicloud": {
      "region": "$${var.region_name}"
    }
  },
  "resource": {
    "huaweicloud_vpc": {
      "test": {
        "name": "$${var.vpc_name}",
        "cidr": "192.168.0.0/16"
      }
    },
    "huaweicloud_vpc_subnet": {
      "test": {
        "vpc_id": "$${huaweicloud_vpc.test.id}",
        "name": "$${var.subnet_name}",
        "cidr": "$${cidrsubnet(huaweicloud_vpc.test.cidr, 4, 1)}",
        "gateway_ip": "$${cidrhost(cidrsubnet(huaweicloud_vpc.test.cidr, 4, 1), 1)}"
      }
    }
  },
  "variable": {
    "region_name": {
      "type": "string"
    },
    "vpc_name": {
      "type": "string"
    },
    "subnet_name": {
      "type": "string"
    }
  }
}
EOT

  vars_body = <<EOT
region_name = "${var.region_name}"
vpc_name    = "${var.vpc_name}"
subnet_name = "${var.subnet_name}"
EOT
}
```

## Argument Reference

The following arguments are supported:

* `region` - (Optional, String, ForceNew) Specifies the region where the RF resource stack is located.  
  If omitted, the provider-level region will be used. Change this parameter will create a new resource.

* `name` - (Required, String, ForceNew) Specifies the name of the resource stack.  

>   The name can contain `1` to `64` characters, only letters, digits and hyphens (-) are allowed.

Reviewer 2: ```suggestion
  The valid length is limited from 1 to 64, and only letters, digits, and hyphens (-) are allowed.
```

Author: ```
The valid length is limited from 1 to 64, and only letters, digits and hyphens (-) are allowed.
```

* `description` - (Optional, String, ForceNew) Specifies the description of the resource stack, which contain maximum of
  255 characters.  
  Change this parameter will create a new resource.

* `agency` - (Optional, String, ForceNew) Specifies the configuration of the agencies authorized to IAC.  
  Change this parameter will create a new resource.
  The [object](#stack_agency) structure is documented below.

* `template_body` - (Optional, String) Specifies the HCL/JSON template content for deployment resources.
  This parameter and `template_uri` are alternative and required if `vars_body` is set.

* `vars_body` - (Optional, String) Specifies the variable content for deployment resources.
  This parameter and `vars_uri` are alternative.

* `template_uri` - (Optional, String) Specifies the OBS address where the HCL/JSON template archive (**.zip** file,
  which contains all resource **.tf.json** script files to be deployed) or **.tf.json** file is located, which describes
  the target status of the deployment resources.

* `vars_uri` - (Optional, String) Specifies the OBS address where the variable (**.tfvars**) file corresponding to the

>   HCL/JSON template is located, which describes the target status of the deployment resources.

Reviewer: s/is located/located

Author: done.

* `enable_auto_rollback` - (Optional, Bool, ForceNew) Specifies whether to enable automatic rollback.  

>   After enabled, the stack will automatically rollback to the last stable state after the deployment fails.

Reviewer 2: ```suggestion
  If enabled, the stack resources will roll back automatically to the last stable state with deployment fails.
```

Author: done.

  The default value is **true**.

>   Change this parameter will create a new resource.

Reviewer 2: ```suggestion
  A new resource will be created by value reversion.
  The new resource will be recreated by value reversion.
  （重建？还是新建）
```

Author: The current description is the general description for provider documentation.

* `enable_deletion_protection` - (Optional, Bool, ForceNew) Specifies whether to enable delete protection.  
  The default value is **false**.
  Change this parameter will create a new resource.

<a name="stack_agency"></a>
The `agency` block supports:

* `name` - (Required, String, ForceNew) Specifies the name of IAM agency authorized to IAC account.  
  Change this parameter will create a new resource.

* `provider_name` - (Required, String, ForceNew) Specifies the name of the provider corresponding to the IAM agency.  
  Change this parameter will create a new resource.

## Attributes Reference

In addition to all arguments above, the following attributes are exported:

> * `id` - The data source ID.

Reviewer: data source?

Author: fixed.

* `status` - The current status of the resource stack.

* `created_at` - The creation time.

* `updated_at` - The latest update time.

## Timeouts

This resource provides the following timeouts configuration options:

* `create` - Default is 20 minutes.
* `update` - Default is 20 minutes.
* `delete` - Default is 5 minutes.

For most HCL templates, the timeout parameters needs to be manually configured by the user to ensure that resources can
be deployed successfully on the RF resource stack, e.g.

```HCL
resource "huaweicloud_rf_stack" "test" {
  ...

  timeouts {
    create = "1h" // Such as the creation of GaussDB instances.
    update = "1h"

>   }

Reviewer 3: if this part is in the template body, add EOT around  it to make the example clear

Author: Actually, it is the timeout for resource stack, not the timeout for template resource.

Reviewer: hah, seems we use tf in tf is a bit difficult for users, better to suggest use obs url.

Author: sure, obs URLs are more clear

}
```

## Import

> Stacks can be imported using their `id`, e.g.

Reviewer 2: ```suggestion
Stacks can be imported with their `id`, e.g.
```

Author: `using` is ok.

```
$ terraform import huaweicloud_rf_stack.test edd2f099-e1ac-4bd0-be32-8b2185620a90
```

Note that the imported state may not be identical to your resource definition, due to some attributes missing from the
API response. The missing attributes include: `agency`, `template_body`, `vars_body`, `template_uri`, `vars_uri`,
`enable_auto_rollback` and `enable_deletion_protection`. It is generally recommended running `terraform plan` after

> importing a stack. You can then decide if changes should be applied to the resource, or the resource definition should

Reviewer 2: ```suggestion
importing a stack. You can keep the resource the same with its definition by choosing any of them to update.
```

Author: done.

---
subcategory: "Application Orchestration Service (AOS)"
---

# huaweicloud_rf_stack

Provides an RF resource stack.

## Example Usage

### Create an RF resource stack with resource deployment (using OBS storage URI)

```HCL
variable "stack_name" {}
variable "agency_name" {}
variable "template_storage_uri" {}

> variable "variable_storage_uri" {}

Reviewer: Nits: how aout template_obs_url and varialbe_obs_url

Author: ok

resource "huaweicloud_rf_stack" "test" {
  name = var.stack_name

  agency {
    name          = var.agency_name
    provider_name = "huaweicloud"
  }

  template_uri = var.template_storage_uri
  vars_uri     = var.variable_storage_uri
```

### Create an RF resource stack with VPC deployment (using template and variable files)

```HCL
variable "stack_name" {}
variable "agency_name" {}
variable "template_storage_full_address" {}

> variable "variable_storage_full_address" {}

Reviewer: Nits: how about template_path and variable_path

Author: ok
